src/targets/gpu/kernels/include/migraphx/kernels/pad.hpp

```diff
@@ -0,0 +1,55 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2022 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_KERNELS_PAD_HPP
+#define MIGRAPHX_GUARD_KERNELS_PAD_HPP
+
+#include <migraphx/kernels/clamp.hpp>
+#include <migraphx/kernels/shape.hpp>
+#include <migraphx/kernels/index.hpp>
+#include <migraphx/kernels/print.hpp>
+
+namespace migraphx {
+
+template <class T, class U, class V, class W, class X, class Y>
+__device__ void
+pad(const T& idx, const U& offsets, const V& bounds, const W& input, X& output, const Y& pad_val)
+{
+    auto output_shape = output.get_shape();
+    idx.global_stride(output_shape.elements(), [&](auto i) {
+        auto multi = output_shape.multi(i);
+
+        for(std::size_t j = 0; j < offsets.size(); j++)
+        {
+            if(multi[j] < offsets[j] or multi[j] >= bounds[j])
+            {
+                output[multi] = pad_val;
+                return;
+            }
+        }
```

Comment:
I dont really understand this. I would do something like this(assuming a lexical comparison):

```cpp
auto input_idx = multi - offsets;
if (multi < offsets or input_idx >= input.get_shape().lens)
    output[multi] = pad_val;
else
    output[multi] = input[input_idx];
```

Does that not work?

Reply:
I got the above to work if I had the following comparison:
```
friend constexpr bool operator<(const array& x, const array& y)
    {
        for(index_int i = 0; i < N; i++)
        {
            if(x[i] >= y[i]) return false;
        }
        return true;
    }
friend constexpr bool operator>=(const array& x, const array& y) { return not( x < y); }

```

Reply:
Actually this should be logically identical to what is in develop currently. Just the operator >= needed to be changed.

Reply:
Yea thinking about this more and lexical ordering wont work. We do want to use product order as we want to check if any index is out of bounds. So the `input_idx >= input.get_shape().lens` check will work without changed to array.hpp, but the `multi < offsets` wont work. We really want to check if `multi - offsets` has negative indices(or rather `input_idx < 0`), which wont work for unsigned integers. We could check that `input_idx > multi`.

Reply:
I've tried the following which still doesn't work:
`if(input_idx > multi or input_idx >= input.get_shape().lens`
I think it's because the `>` overload is just doing `y < x`. If I have a simple example like this:
```
pads:
0, 0, 1, 1, 0, 0, 1, 1 // pad before and after on last 2 dims
offsets:
0, 0, 1, 1
multi:
0, 0, 0, 0
input_idx:
0, 0, 4294967295, 4294967295
input.get_shapes().lens:
1, 1, 2, 2
```
What happens is that it will iterate through the first 2 dimensions and identify that `0 < 0` is false, so the `input_idx > multi` condition fails.

Reply:
Yea we should probably just do:

```cpp
bool use_pad = false;
for(index_int i = 0; i < multi.size(); i++)
{
    if (input_idx[i] > multi[i] or input_idx[i] >= input.get_shape().lens[i])
    {
        use_pad = true;
        break;
    }   
}
```

Although it would be cleaner to use `range(multi.size())` with `any_of` instead.